appcelerator/titanium_mobile
# refactor(android)(9_3_X): honor Android 11 "package visibility"

**JIRA:**
https://jira.appcelerator.org/browse/TIMOB-28050

**Summary:**
- Now auto-injects `<queries/>` entry if `Ti.UI.createEmailDialog()` API is used.
  * Needed to make [EmailDialog.isSupport()](https://docs.appcelerator.com/platform/latest/#!/api/Titanium.UI.EmailDialog-method-isSupported) work.
- Modified `Ti.Platform.openURL()` to not fail if `<queries/>` entry is missing in "tiapp.xml" file.
  * The `<queries/>` entry is now only required for `Ti.Platform.canOpenURL()`.
- Updated docs for `Ti.Platform.canOpenURL()` with `<queries/>` URL scheme examples.

---
**E-Mail Dialog Test:**
1. Create a "Classic" project with the below "tiapp.xml" setting and "app.js" code.
2. Build and run on Android 11.
3. Tap on the "Send E-Mail" button.
4. Verify the device's e-mail app appears.
5. Open Finder/WindowsExplorer.
6. Go to folder: `./build/android/app/build/intermediates/merged_manifests/debug`
7. Verify "AndroidManifest.xml" contains a `<queries/>` block with mimeType set to "message/rfc822".

tiapp.xml
```xml
<ti:app>
	<android>
		<manifest>
			<uses-sdk android:targetSdkVersion="30"/>
		</manifest>
	</android>
</ti:app>
```

app.js
```javascript
var window = Ti.UI.createWindow();
var button = Ti.UI.createButton({ title: "Send E-Mail" });
button.addEventListener("click", function() {
	var dialog = Ti.UI.createEmailDialog();
	if (dialog.isSupported()) {
		dialog.subject = "E-Mail Test";
		dialog.toRecipients = ["john.doe@domain.com", "jane.doe@domain.com"];
		dialog.messageBody = "This is the e-mail's body.\nThis is the second line.";
		dialog.open();
	} else {
		alert("E-mail app not configured on this device.");
	}
});
window.add(button);
window.open();
```

---
**canOpenURL() Test:**
_(This test is based on PR #10881 .)_
1. Create a "Classic" project with the below "tiapp.xml" settings and "app.js" code.
2. Build and run on Android 11.
3. Tap on every "canOpenURL" button and verify they all display an alert displaying `true`. _(The "tel" and "geo" URLs will return `false` if a phone or maps app is not installed respectively.)_
4. Delete the entire `<queries/>` block from the "tiapp.xml" file.
5. Rebuild and run on Android 11.
6. Tap on every "canOpenURL" button and verify they all display `false` in the alert. _(Note that they will still return `true` on Android 10 and older.)_
7. Tap on every "openURL" button and verify that they display the URL's contents. _(See PR #10881 for what they should display.)_

tiapp.xml
```xml
<ti:app>
	<android>
		<manifest>
			<uses-sdk android:targetSdkVersion="30"/>
			<queries>
				<intent>
					<action android:name="android.intent.action.VIEW"/>
					<data android:scheme="geo"/>
				</intent>
				<intent>
					<action android:name="android.intent.action.VIEW"/>
					<data android:scheme="https"/>
				</intent>
				<intent>
					<action android:name="android.intent.action.VIEW"/>
					<data android:mimeType="image/*"/>
				</intent>
				<intent>
					<action android:name="android.intent.action.VIEW"/>
					<data android:scheme="mailto"/>
				</intent>
				<intent>
					<action android:name="android.intent.action.VIEW"/>
					<data android:scheme="tel"/>
				</intent>
			</queries>
		</manifest>
	</android>
</ti:app>
```

app.js
```javascript
function createTableRowForUrl(url) {
	var row = Ti.UI.createTableViewRow({
		layout: "horizontal",
	});
	row.add(Ti.UI.createLabel({
		text: url,
		top: "10dp",
		left: "10dp",
		right: "10dp",
		width: Ti.UI.FILL,
		height: Ti.UI.SIZE,
	}));
	var canOpenButton = Ti.UI.createButton({
		title: "canOpenURL()",
		top: "10dp",
		bottom: "10dp",
		left: "10dp",
		width: Ti.UI.SIZE,
		height: Ti.UI.SIZE,
	});
	canOpenButton.addEventListener("click", function() {
		var canOpen = Ti.Platform.canOpenURL(url);
		Ti.API.info("@@@ canOpenURL() returned '" + canOpen + "' for URL: " + url);
		alert("canOpenURL() returned:\n" + canOpen);
	});
	row.add(canOpenButton);
	var openButton = Ti.UI.createButton({
		title: "openURL()",
		top: "10dp",
		bottom: "10dp",
		left: "10dp",
		width: Ti.UI.SIZE,
		height: Ti.UI.SIZE,
	});
	openButton.addEventListener("click", function() {
		var wasOpened = Ti.Platform.openURL(url);
		Ti.API.info("@@@ openURL() returned '" + wasOpened + "' for URL: " + url);
	});
	row.add(openButton);
	return row;
}

var tableRows = [];
tableRows.push(createTableRowForUrl("http://www.appcelerator.com"));
tableRows.push(createTableRowForUrl("https://www.appcelerator.com"));
tableRows.push(createTableRowForUrl("mailto:user@domain.com"));
tableRows.push(createTableRowForUrl("mailto:user@domain.com?cc=johndoe@domain.com,janedoe@domain.com&bcc=bsmith@domain.com&subject=The%20Subject&body=This%20is%20a%20test.%0AHello%20World!"));
tableRows.push(createTableRowForUrl("tel:123-456-7890"));
tableRows.push(createTableRowForUrl("geo:37.774,-122.431?z=12"));  // San Francisco
tableRows.push(createTableRowForUrl("badscheme://i.dont.exist"));
if (OS_ANDROID) {
	var sourceFile = Ti.Filesystem.getFile(Ti.Filesystem.resourcesDirectory, "assets/images/tab1@2x.png");
	var externalFilePath = Ti.Filesystem.applicationDataDirectory + "Copy of " + sourceFile.name;
	sourceFile.copy(externalFilePath);
	tableRows.push(createTableRowForUrl("assets/images/tab1.png"));
	tableRows.push(createTableRowForUrl("/assets/images/tab1.png"));
	tableRows.push(createTableRowForUrl("././assets/images/.././images/tab1.png"));
	tableRows.push(createTableRowForUrl(sourceFile.nativePath));
	tableRows.push(createTableRowForUrl(externalFilePath));
	tableRows.push(createTableRowForUrl(Ti.Filesystem.getFile(externalFilePath).nativePath));
	tableRows.push(createTableRowForUrl("content://" + Ti.App.id + ".tifileprovider/assets/Resources/assets/images/tab1.png"));
	tableRows.push(createTableRowForUrl(Ti.Filesystem.resourcesDirectory + "assets/images/tab1.png"));
	tableRows.push(createTableRowForUrl("file:///android_asset/Resources/assets/images/tab1.png"));
	tableRows.push(createTableRowForUrl("android.resource://" + Ti.App.id + "/drawable/appicon"));
}

var window = Ti.UI.createWindow({
	extendSafeArea: false,
	fullscreen: true,
	theme: "Theme.AppCompat.NoTitleBar",
});
var tableView = Ti.UI.createTableView({
	data: tableRows,
});
window.add(tableView);
window.open();
```
